From d519eac2257ecf7f28d65b4d91f726f243fc9a54 Mon Sep 17 00:00:00 2001
From: Jason Jae Hyun Koh <jason.koh@motorola.com>
Date: Tue, 19 Jan 2016 13:29:32 -0600
Subject: [PATCH 633/959] IKSWM-17218 drivers:misc:stml0xx: handle queue full
 flag

Let the driver log if the queue full flag is set in the non-wake irq status,
instead of letting the sensorhub log by default and trigger an AP wake.

Change-Id: I05af7ae9c8e03d8c7a041d0050b87c2d819f0bdc
Signed-off-by: Jason Jae Hyun Koh <jason.koh@motorola.com>
Reviewed-on: http://gerrit.mot.com/814363
SME-Granted: SME Approvals Granted
SLTApproved: Slta Waiver <sltawvr@motorola.com>
Tested-by: Jira Key <jirakey@motorola.com>
Reviewed-by: George Standish <george.standish@motorola.com>
Reviewed-by: Daniel Moore <dtmoore@motorola.com>
Reviewed-by: Eric Hefner <ehefner@motorola.com>
Reviewed-by: Christopher Fries <cfries@motorola.com>
Submit-Approved: Jira Key <jirakey@motorola.com>
---
 drivers/misc/stml0xx/stml0xx_irq.c | 10 +++++++---
 include/uapi/linux/stml0xx.h       |  6 ++++--
 2 files changed, 11 insertions(+), 5 deletions(-)

diff --git a/drivers/misc/stml0xx/stml0xx_irq.c b/drivers/misc/stml0xx/stml0xx_irq.c
index 94b037f5891..941faf22471 100644
--- a/drivers/misc/stml0xx/stml0xx_irq.c
+++ b/drivers/misc/stml0xx/stml0xx_irq.c
@@ -284,10 +284,14 @@ void stml0xx_irq_work_func(struct work_struct *work)
 	    (buf[IRQ_IDX_STATUS_MED] << 8) | buf[IRQ_IDX_STATUS_LO];
 
 	/* Check for streaming sensors */
-	if (((stm_ws->flags & IRQ_WORK_FLAG_DISCARD_SENSOR_QUEUE) == 0) &&
-	     (irq_status & (M_ACCEL | M_ACCEL2 | M_GYRO | M_UNCALIB_GYRO))) {
-		stml0xx_process_stream_sensor_queue(buf, stm_ws->ts_ns);
+	if ((stm_ws->flags & IRQ_WORK_FLAG_DISCARD_SENSOR_QUEUE) == 0) {
+		if (irq_status & M_QUEUE_OVERFLOW)
+			dev_err(&stml0xx_misc_data->spi->dev,
+				"Streaming sensor queue full");
+		if (irq_status & (M_ACCEL | M_ACCEL2 | M_GYRO | M_UNCALIB_GYRO))
+			stml0xx_process_stream_sensor_queue(buf, stm_ws->ts_ns);
 	}
+
 	if (irq_status & M_ALS) {
 		stml0xx_as_data_buffer_write(
 			ps_stml0xx,
diff --git a/include/uapi/linux/stml0xx.h b/include/uapi/linux/stml0xx.h
index 6a769fb02f6..ec2dcf09268 100644
--- a/include/uapi/linux/stml0xx.h
+++ b/include/uapi/linux/stml0xx.h
@@ -188,6 +188,7 @@
 #define M_UNCALIB_MAG           0x010000
 #define M_ACCEL2                0x020000
 #define M_QUAT_9AXIS            0x040000
+#define M_QUEUE_OVERFLOW        0x080000
 
 /* wake sensor status */
 #define M_DOCK                  0x000001
@@ -351,8 +352,9 @@ enum lowpower_mode {
 enum sh_log_level {
 	SH_LOG_DISABLE,
 	SH_LOG_ERROR,
-	SH_LOG_VERBOSE,
-	SH_LOG_DEBUG
+	SH_LOG_WARN,
+	SH_LOG_DEBUG,
+	SH_LOG_VERBOSE
 };
 
 struct stm_response {
-- 
2.11.0

